### PR TITLE
Drop python3.7

### LINKED ISSUE #966: Drop Python3.7
Starting from September 2021, Python3.7 will only be receiving security fixes ([PEP 537](https://www.python.org/dev/peps/pep-0537/)). This being the case, Python3.7 is gradually being phased out by many scientific software dependencies we rely upon.

To borrow a few links from the `rioxarray` project (https://github.com/corteva/rioxarray/issues/451):
Pandas dropping Python3.7: https://github.com/pandas-dev/pandas/issues/41678
NumPy dropping Python3.7: https://numpy.org/neps/nep-0029-deprecation_policy.html

Once Xarray drops Python3.7, we should follow suit. 

In the meantime, here are the new features that will be allowed for **[Python3.8](https://www.python.org/dev/peps/pep-0569/#id8)**.

EDIT: Python3.8 will give us the ["Walrus Operator"](https://www.python.org/dev/peps/pep-0572/)!

### PR DESCRIPTION
**Pull request by Zeitsperre**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #966 
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.
- [x] `bumpversion patch` has been called on this branch
- [ ] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?

* Updated Black to 22.1.0 and enforces Python3.8 conventions
* Corrected various docstring, grammar, and type hinting errors
* Added more reference tags in docstrings
* bumped version to 0.33.1-beta
* Updated package metadata and `tox` to drop Python3.7
* Updated History
* Dropped Python3.7 in CI builds

### Does this PR introduce a breaking change?
Yes. Code conventions now use Python3.8. This doesn't change much for existing code, beyond whitespace policies.

### CODE REVIEW

File: docs/notebooks/cli.ipynb [hunk 1]
```diff
@@ -262,7 +262,11 @@
    "source": [
     "# Create an output file with just the flag value and no aggregation (dims=None)\n",
     "\n",
-    "!xclim -i example_data.nc -o flag_output.nc dataflags -d none"
+    "!xclim -i example_data.nc -o flag_output.nc dataflags -d none\n",
+    "\n",
+    "# Need to wait until the file is written\n",
+    "\n",
+    "!sleep 2s"
```
Comment (aulemahal) [inline docs/notebooks/cli.ipynb:269]: Does the command returns before the file is written? Weird!?

**Review by aulemahal — APPROVED** [review]

File: docs/notebooks/cli.ipynb (hunk 1 above)
Comment (Zeitsperre) [reply docs/notebooks/cli.ipynb:269]: It was happening to me, so I added it in case. Weird behaviour for sure.

**Comment by Zeitsperre** [comment]
In the past week, the following libraries have been released:
* black is now stable at 22.1.0 (and is showing some strange behaviour with existing checks)
* xarray 0.21 (dropping Python3.7)

It appears that there are some edge case that the CLI is tripping on with the newest xarray. The work continues...

**Review by bzah — APPROVED** [review]
You may want to update python version used in ` .github/workflows/` as well.
Other than that it LGTM, thanks!

**Review by juliettelavoie — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -21,29 +17,34 @@ repos:
...
+    hooks:
+      - id: nbqa-black
+      - id: nbqa-pyupgrade
+        args: [--py38-plus]
+      - id: nbqa-isort
+#   - repo: https://github.com/kynan/nbstripout
+#     rev: 0.5.0
+#     hooks:
+#       - id: nbstripout
+#         files: ".ipynb"
```
Comment (aulemahal) [inline .pre-commit-config.yaml:47]: What happened here?

**Review by aulemahal — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Zeitsperre) [reply .pre-commit-config.yaml:47]: `nbstripout` was causing issues with `black-jupyter`, due to an existing bug/feature (https://github.com/kynan/nbstripout/issues/121). I turned it off for the time being. We can probably re-enable it once the notebooks are in working order again.

**Review by aulemahal — APPROVED** [review]

**Pull request merged**